src/python/pants/backend/terraform/lint/trivy/rules.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Any
+
+from pants.backend.terraform.dependencies import (
+    prepare_terraform_invocation,
+    terraform_fieldset_to_init_request,
+)
+from pants.backend.terraform.dependency_inference import (
+    TerraformDeploymentInvocationFilesRequest,
+    get_terraform_backend_and_vars,
+)
+from pants.backend.terraform.target_types import (
+    TerraformDeploymentFieldSet,
+    TerraformDeploymentTarget,
+    TerraformFieldSet,
+    TerraformModuleTarget,
+)
+from pants.backend.tools.trivy.rules import RunTrivyRequest, run_trivy
+from pants.backend.tools.trivy.subsystem import SkipTrivyField, Trivy
+from pants.core.goals.lint import LintResult, LintTargetsRequest
+from pants.core.util_rules.partitions import PartitionerType
+from pants.core.util_rules.source_files import SourceFilesRequest, determine_source_files
+from pants.engine.internals.native_engine import MergeDigests
+from pants.engine.intrinsics import merge_digests
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import collect_rules, implicitly, rule
+from pants.engine.target import FieldSet, SourcesField, Target
+from pants.util.logging import LogLevel
+
+
+class TrivyLintTerraformRequest(LintTargetsRequest, metaclass=ABCMeta):
+    tool_subsystem = Trivy
+
+
+@dataclass(frozen=True)
+class TrivyTerraformFieldSet(FieldSet, metaclass=ABCMeta):
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipTrivyField).value
+
+
+@dataclass(frozen=True)
+class RunTrivyOnTerraformRequest:
+    field_set: TrivyTerraformFieldSet
+
+
+@rule
+async def run_trivy_on_terraform(req: RunTrivyOnTerraformRequest) -> FallibleProcessResult:
+    fs = req.field_set
+    # Each subclass of TrivyTerraformFieldSet is a subclass of either TerraformDeploymentFieldSet or TerraformFieldSet
+    tf = await prepare_terraform_invocation(terraform_fieldset_to_init_request(fs))  # type: ignore
+    command_args = []
+
+    if isinstance(fs, TerraformDeploymentFieldSet):
```

Comment:
@lilatomic I'm zombie threading here - but what was the purpose of this check? It's failing mypy in Python 3.14 - as a `TrivyTerraformFieldSet` can't be a `TerraformDeploymentFieldSet`

```
src/python/pants/backend/terraform/lint/trivy/rules.py:58:19: error: Subclass of "TrivyTerraformFieldSet" and "TerraformDeploymentFieldSet" cannot exist: would have incompatible method signatures  [unreachable]
        if isinstance(fs, TerraformDeploymentFieldSet):
                      ^~
src/python/pants/backend/terraform/lint/trivy/rules.py:60:9: error: Statement is unreachable  [unreachable]
            invocation_files = await get_terraform_backend_and_vars(
```

Reply:
Deployments and Modules are different in that Deployments have Backend configs and Var files.
I think the warning is correct in that TrivyTerraformFieldSet doesn't specify that it is a sealed class (that's the term in Kotlin). Looking into the Mypy error, there might be something about `TrivyLintTerraformDeploymentFieldSet` having a conflict in a method or something?

Reply:
Maybe. It's typed as one fieldset without an overlap, so it's kinda weird to see that check here. I guess that's more an artifact of how Pants data gets passed around?